File: neetcode150/easy/cpp/242-Valid-Anagram.cpp
#include <map>
#include <string>
using namespace std;

class Solution {
public:
    // Two map solution:
    // Store the frequency of each letter in a map and
    // compare the frequencies of the two strings.
    bool isAnagram(string s, string t) {
        if(s.size() != t.size()){
            return false;
        }
        map<char, int> freq_s;    
        map<char, int> freq_t;    
        // Add frequencies of string s
        for(int i = 0; i < s.size(); i++){
            if(freq_s.find(s[i]) == freq_s.end()) {
                freq_s.insert(make_pair(s[i], 1));
            } else{
                freq_s[s[i]]++;
            }
        }
        // Add frequencies of string t
        for(int j = 0; j < t.size(); j++){
            if(freq_t.find(t[j]) == freq_t.end()) {
                freq_t.insert(make_pair(t[j], 1));
            } else{
                freq_t[t[j]]++;
            }
        }
        // Compare frequencies of the two strings
        for(int k = 0; k < s.size(); k++){
            if(freq_s[s[k]] != freq_t[s[k]]){
                return false;
            }
        } 
        return true;
    }
};

class Solution {
public:
    // Single array solution:
    // Create an array of all 26 letters and increment when 
    // a character is found in string s and decrement when 
    // a character is found in string t.
    bool isAnagram(string s, string t) {
        if(s.size() != t.size()){
            return false;
        }
        int count[26] = {};
        for(int i = 0; i < s.size(); i++){
            count[s[i] - 'a']++;
            count[t[i] - 'a']--;
        }
        
        for(int i : count){
            if(i != 0){
                return false;
            }
        }
        return true;
    }
};

File: neetcode150/easy/242-Valid-Anagram.py
#!/usr/bin/python3

class Solution:
  def isAnagram(self, s: str, t: str) -> bool:
    # Check if the two strings are equivalent lengths
    if len(s) != len(t):
      return False

    # Initialize two dictionaries to store the frequency 
    # of each letter in the given strings
    h1 = {}
    h2 = {}

    # Add the letters of string s to the dictionary h1
    for l in s:
      if l in h1:
        h1[l] += 1
      else:
        h1[l] = 1
  
    # Add the letters of string t to the dictionary h2
    for l in t:
      if l in h2:
        h2[l] += 1
      else:
        h2[l] = 1
    
    # Check if the two dictinaries are equivalent
    return h1 == h2

  # Solution inspired by C++ array solution
  def isAnagram(self, s: str, t: str) -> bool:
    # Check if the two strings are the same length
    if len(s) != len(t):
      return False

    # Create an array of the count of each character
    count = [0] * 26
    for i in range(len(s)):
      # Increment if in s 
      count[ord(s[i]) - ord('a')] += 1
      # Decrement if in t
      count[ord(t[i]) - ord('a')] -= 1
    
    # If any of the counts != 0 then we know the strings 
    # are NOT anagrams
    for n in count:
      if n != 0:
        return False
    return True
